src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,730 @@
+        that->_window = wil::unique_hwnd(window);
+        SetWindowLongPtr(that->_window.get(), GWLP_USERDATA, reinterpret_cast<LONG_PTR>(that));
+    }
+    else if (WindowEmperor* that = GetThisFromHandle(window))
+    {
+        return that->MessageHandler(message, wparam, lparam);
+    }
+
+    return DefWindowProc(window, message, wparam, lparam);
+}
+void WindowEmperor::_createMessageWindow()
+{
+    WNDCLASS wc{};
+    wc.hCursor = LoadCursor(nullptr, IDC_ARROW);
+    wc.hInstance = reinterpret_cast<HINSTANCE>(&__ImageBase);
+    wc.lpszClassName = TERMINAL_MESSAGE_CLASS_NAME;
+    wc.style = CS_HREDRAW | CS_VREDRAW;
+    wc.lpfnWndProc = MessageWndProc;
+    wc.hIcon = LoadIconW(wc.hInstance, MAKEINTRESOURCEW(IDI_APPICON));
+    RegisterClass(&wc);
+    WINRT_ASSERT(!_window);
+
+    WINRT_VERIFY(CreateWindow(wc.lpszClassName,
+                              L"Windows Terminal",
+                              0,
+                              CW_USEDEFAULT,
+                              CW_USEDEFAULT,
+                              CW_USEDEFAULT,
+                              CW_USEDEFAULT,
+                              HWND_MESSAGE,
+                              nullptr,
+                              wc.hInstance,
+                              this));
+}
+
+LRESULT WindowEmperor::MessageHandler(UINT const message, WPARAM const wParam, LPARAM const lParam) noexcept
+{
+    switch (message)
+    {
+    case WM_HOTKEY:
+    {
+        _hotkeyPressed(static_cast<long>(wParam));
+        return 0;
+    }
+    case CM_NOTIFY_FROM_NOTIFICATION_AREA:
+    {
+        switch (LOWORD(lParam))
+        {
+        case NIN_SELECT:
+        case NIN_KEYSELECT:
+        {
+            _notificationIcon->NotificationIconPressed();
+            return 0;
+        }
+        case WM_CONTEXTMENU:
+        {
+            const til::point eventPoint{ GET_X_LPARAM(wParam), GET_Y_LPARAM(wParam) };
+            _notificationIcon->ShowContextMenu(eventPoint, _manager.GetPeasantInfos());
+            return 0;
+        }
+        }
+        break;
+    }
+    case WM_MENUCOMMAND:
+    {
+        _notificationIcon->MenuItemSelected((HMENU)lParam, (UINT)wParam);
+        return 0;
+    }
+    default:
+    {
+        // We'll want to receive this message when explorer.exe restarts
+        // so that we can re-add our icon to the notification area.
+        // This unfortunately isn't a switch case because we register the
+        // message at runtime.
+        if (message == WM_TASKBARCREATED)
+        {
+            _notificationIcon->ReAddNotificationIcon();
+            return 0;
+        }
+    }
+    }
+    return DefWindowProc(_window.get(), message, wParam, lParam);
+}
+
+winrt::fire_and_forget WindowEmperor::_close()
+{
+    // Important! Switch back to the main thread for the emperor. That way, the
+    // quit will go to the emperor's message pump.
+    co_await wil::resume_foreground(_dispatcher);
+    PostQuitMessage(0);

```

Comment:
Shouldn't this post a `WM_CLOSE` instead of directly killing the thread to be conforming to the Win32 style? It would require the addition of a `case WM_CLOSE` above though. It's unfortunate we don't have the UI thread ID here, because then we could use `PostThreadMessageW`.

Reply:
I thought that `WM_QUIT` was the idiomatic way here.

> `GetMessage` will return 0 when it pulls a `WM_QUIT` out

So that's what gets us out of the loop in the main thread, in `WaitForWindows`. I don't think it immediately terminates the thread, it just appends it to the main thread's message queue